File: Assignment_1/apriori_algo_sim.py
import pandas as pd
from itertools import combinations
from collections import defaultdict

def read_dataset(filepath):
    with open(filepath, 'r') as file:
        lines = file.readlines()
    
    transactions = []
    for line in lines:
        transaction = line.strip().split()
        transactions.append(transaction)
    
    print("File read complete")
    return transactions

def find_frequent_1_itemsets(transactions, min_sup):
    item_count = defaultdict(int)
    for transaction in transactions:
        for item in set(transaction):  # using set to avoid duplicates in one transaction
            item_count[item] += 1
    
    L1 = []
    for item, count in item_count.items():
        if count >= min_sup:
            L1.append(frozenset([item]))
    
    return sorted(L1, key=lambda x: sorted(x))

def has_infrequent_subset(candidate, prev_frequent_itemsets):
    for item in candidate:
        subset = candidate - frozenset([item])
        if subset not in prev_frequent_itemsets:
            return True
    return False

def apriori_gen(Lk_minus_1, k):
    candidates = []
    Lk_minus_1 = list(Lk_minus_1)

    for i in range(len(Lk_minus_1)):
        for j in range(i + 1, len(Lk_minus_1)):
            l1 = list(Lk_minus_1[i])
            l2 = list(Lk_minus_1[j])

            if l1[:k-1] == l2[:k-1]:
                candidate = Lk_minus_1[i] | Lk_minus_1[j]
                if not has_infrequent_subset(candidate, Lk_minus_1):
                    candidates.append(candidate)
    return candidates

def count_support(transactions, itemset):
    count = 0
    for transaction in transactions:
        if itemset.issubset(transaction):
            count += 1
    return count

def apriori(transactions, min_sup):
    transactions = list(map(set, transactions))
    L = []
    
    L1 = find_frequent_1_itemsets(transactions, min_sup)
    L.append(L1)
    k = 2
    Lk_minus_1 = set(L1)

    while Lk_minus_1:
        Ck = apriori_gen(Lk_minus_1, k - 1)
        Lk = []
        for candidate in Ck:
            if count_support(transactions, candidate) >= min_sup:
                Lk.append(candidate)
        if not Lk:
            break
        L.append(Lk)
        Lk_minus_1 = set(Lk)
        k += 1

        print(f"Candidates for {k}-itemsets: {Ck}")
        print(f"Frequent {k}-itemsets: {Lk}")

    return L

def print_itemsets(L):
    for k, itemsets in enumerate(L):
        print(f"Frequent {k+1}-itemsets:")
        for itemset in itemsets:
            print(sorted(itemset))
        print()

# Example usage
if __name__ == "__main__":
    filepath = "Datasets/retail.dat.txt"
    transactions = read_dataset(filepath)
    min_support = 2
    L = apriori(transactions, min_support)
    print_itemsets(L)

File: Assignment_1/apriori_algo_sim.cpp
#include <bits/stdc++.h>

using namespace std;
#define fast_cin() ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL)

#define     all(x)      x.begin(), x.end() 

typedef long long int ll;



typedef set<string> ItemSet;
typedef vector<ItemSet> List_of_ItemSets;
typedef vector<vector<string>> Transactions;


ostream& operator<<(ostream& os ,ItemSet& itemSet){
    os<<"{";

    for(auto it=itemSet.begin(); it!=itemSet.end(); ++it){
        os<< (*it) <<" ";
    }

    os <<"}";

    return os;
}


ostream& operator<<(ostream& os, List_of_ItemSets& L){

    os<<"[";

    for(ItemSet itemSet : L){
        os<<itemSet;
        os<<endl;
    }

    os<<"]";

    return os;
}


List_of_ItemSets find_frequent_1_itemSet(vector<vector<string>> &D,int min_sup){
    ItemSet set;

    for(int i=0; i<D.size(); i++){
        for(string s: D[i]){
            set.insert(s);
        }
    }

    List_of_ItemSets L1;

    for(string s : set){
        int count = 0;

        // cout<<s<<": "<<endl;
        for(int i=0; i<D.size(); i++){
            if(find(D[i].begin(), D[i].end() ,s) != D[i].end()){
                count++;
            }
        }

        // cout<<count<<endl;

        if(count >= min_sup){
            ItemSet itemSet;
            itemSet.insert(s);
            L1.push_back(itemSet);
        }
    }

    // for(string s : set){
    //     cout<<s<<endl;
    // }

    sort(all(L1));

    return L1;
}


ItemSet join(ItemSet &l1, ItemSet &l2, int k_1) {
    auto it1 = l1.begin();
    auto it2 = l2.begin();

    ItemSet c;

    for (int i = 0; i < k_1 - 1; ++i, ++it1, ++it2) {
        if (*it1 != *it2) {
            return {}; 
        }
        c.insert(*it1); 
    }

    c.insert(*it1);
    c.insert(*it2);

    return c;
}



bool has_infrequent_subset(ItemSet c, List_of_ItemSets Lk_1){
    for(auto it=c.begin();it!=c.end(); ++it){
        ItemSet subset = c;
        subset.erase(*it);


        if (find(Lk_1.begin(), Lk_1.end(), subset) == Lk_1.end()) {
            return true; // found an infrequent subset
        }
    }

    return false;
}


List_of_ItemSets apriori_gen(List_of_ItemSets Lk_1, int k_1) {
    List_of_ItemSets Ck;


    for (size_t i = 0; i < Lk_1.size(); ++i) {
        for (size_t j = i + 1; j < Lk_1.size(); ++j) {
            ItemSet l1 = Lk_1[i];
            ItemSet l2 = Lk_1[j];

            auto it1 = l1.begin();
            auto it2 = l2.begin();

            bool mismatch = false;
            for (int k = 0; k < k_1 - 1; ++k, ++it1, ++it2) {
                if (*it1 != *it2) {
                    mismatch = true;
                    break;
                }
            }

            if (mismatch) continue;

            // Check last element to avoid duplicate joins

            auto last1 = *prev(l1.end());
            auto last2 = *prev(l2.end());

            if (last1 >= last2) continue;

            ItemSet c = join(l1, l2, k_1);

            if (!has_infrequent_subset(c, Lk_1)) {
                Ck.push_back(c);
            }
        }
    }

    return Ck;
}

int Counter(Transactions D, ItemSet c){
    int count = 0;
    
    for(int i=0; i<D.size(); i++){
        bool not_found = false;

        for(string s : c){

            if(find(D[i].begin(), D[i].end(), s) == D[i].end()){
                not_found = true;
                break;
            }
        }

        if(!not_found){
            count++;
        }
    }

    return count;
}

Transactions DatasetCreation(void){
    string filepath = "Datasets/retail.dat.txt";
    Transactions D;
    ifstream file(filepath);
    string line;

    while (getline(file, line)) {
        istringstream iss(line);
        vector<string> transaction;
        string item;

        while (iss >> item) {
            transaction.push_back(item);
        }

        D.push_back(transaction);
    }

    file.close();
    cout<< "File read complete"<<endl;
    return D;
}

void solve(void){

    // a database of transaction
    // Transactions D = {
    //     {"I1", "I2", "I5"},
    //     {"I2", "I4"},
    //     {"I2", "I3"},
    //     {"I1", "I2", "I4"},
    //     {"I1", "I3"},
    //     {"I2", "I3"},
    //     {"I1", "I3"},
    //     {"I1", "I2", "I3", "I5"},
    //     {"I1", "I2", "I3"}
    // }; 


    Transactions D = DatasetCreation();

    // const uint32_t min_sup = 2; //min support count

    // List_of_ItemSets L1 = find_frequent_1_itemSet(D, min_sup);

    // for(ItemSet itemSet : L1){
    //     cout<< itemSet << endl;
    // }

    // List_of_ItemSets Lk_1 = L1;

    // for(int k=2;k<4 ;k++){
    //     List_of_ItemSets Ck = apriori_gen(Lk_1, k-1);
    
    // if(Ck.size() == 0){
    //     cout<<"break"<<endl;
    //     break;
    // }

    // cout<<"Ck"<<endl;
    // cout<<Ck<<endl;
    
    // List_of_ItemSets Lk;

    // for(ItemSet itemSet : Ck){
    //     int count = Counter(D, itemSet);

    //     if(count >= min_sup) {
    //         Lk.push_back(itemSet);
    //     }
    // }

    // cout<<"Lk"<<endl;
    // cout<<Lk<<endl;

    // Lk_1 = Lk;

    // }

}

int main() {
fast_cin();


solve();



return 0;
}